pytorch_lightning/utilities/fetching.py

```diff
@@ -228,44 +228,33 @@ def wait(self) -> None:
         """Hook to override to indicate the `DataFetcher` to wait for an event."""
 
     def prefetching(self) -> None:
+        iterator = self.dataloader_iter
         for _ in range(self.prefetch_batches):
             try:
-                self._fetch_next_batch()
+                self._fetch_next_batch(iterator)
             except StopIteration:
                 break
 
-    def fetching_function(self) -> Optional[Tuple[Any, bool]]:
-        if self.done:
-            while self.batches:
-                return self._get_queued_batch()
-            raise StopIteration
+    def fetching_function(self) -> Tuple[Any, bool]:
+        if self.batches:
+            batch = self.batches.pop(0)
         else:
+            # empty iterator, no prefetching done
+            raise StopIteration
+        if not self.done:
+            assert self.dataloader_iter is not None
             try:
-                yield_batch = self.batches.pop(0)
-                self._fetch_next_batch()
-                # wait for batch to be available.
-                self.wait()
-                # yield last and has next
-                return self.move_to_device(yield_batch), False
+                self._fetch_next_batch(self.dataloader_iter)
             except StopIteration:
-                self.batches.insert(0, yield_batch)
                 self.done = True
-                return self._get_queued_batch()
-
-            except IndexError:
-                raise StopIteration
+        self.wait()
+        return self.move_to_device(batch), len(self.batches) == 0
```

Comment:
How do you get this batch in case of no pre-fetching?

Reply:
I'll follow-up in your other comment which I think asks the same thing